cluster/manifests.go

```diff
@@ -1,68 +1,47 @@
 package cluster
 
 import (
+	"bytes"
 	"fmt"
-	"io/ioutil"
-	"os"
-	"path/filepath"
 
 	"github.com/weaveworks/flux"
 	"github.com/weaveworks/flux/image"
 	"github.com/weaveworks/flux/policy"
 	"github.com/weaveworks/flux/resource"
 )
 
-type ManifestError struct {
-	error
-}
-
-func ErrResourceNotFound(name string) error {
-	return ManifestError{fmt.Errorf("manifest for resource %s not found under manifests path", name)}
-}
-
-// Manifests represents how a set of files are used as definitions of
+// Manifests represents a set of files containing definitions of
 // resources, e.g., in Kubernetes, YAML files describing Kubernetes
 // resources.
 type Manifests interface {
-	// Update the image in a manifest's bytes to that given
-	UpdateImage(def []byte, resourceID flux.ResourceID, container string, newImageID image.Ref) ([]byte, error)
 	// Load all the resource manifests under the paths
 	// given. `baseDir` is used to relativise the paths, which are
 	// supplied as absolute paths to directories or files; at least
 	// one path should be supplied, even if it is the same as `baseDir`.
 	LoadManifests(baseDir string, paths []string) (map[string]resource.Resource, error)
-	// UpdatePolicies modifies a manifest to apply the policy update specified
-	UpdatePolicies([]byte, flux.ResourceID, policy.Update) ([]byte, error)
+	// ParseManifest parses the content of a manifest and its source location into resources
+	ParseManifest(def []byte, source string) (map[string]resource.Resource, error)
+	// Set the image of a container in a manifest's bytes to that given
+	SetWorkloadContainerImage(def []byte, resourceID flux.ResourceID, container string, newImageID image.Ref) ([]byte, error)
+	// UpdateWorkloadPolicies modifies a manifest to apply the policy update specified
+	UpdateWorkloadPolicies(def []byte, id flux.ResourceID, update policy.Update) ([]byte, error)
+	// CreateManifestPatch obtains a patch between the original and modified manifests
+	CreateManifestPatch(originalManifests, modifiedManifests []byte, originalSource, modifiedSource string) ([]byte, error)
+	// ApplyManifestPatch applies a manifest patch (obtained with CreateManifestDiff) returned the patched manifests
+	ApplyManifestPatch(originalManifests, patchManifests []byte, originalSource, patchSource string) ([]byte, error)
 }
 
-// UpdateManifest looks for the manifest for the identified resource,
-// reads its contents, applies f(contents), and writes the results
-// back to the file.
-func UpdateManifest(m Manifests, root string, paths []string, id flux.ResourceID, f func(manifest []byte) ([]byte, error)) error {
-	resources, err := m.LoadManifests(root, paths)
-	if err != nil {
-		return err
-	}
-
-	resource, ok := resources[id.String()]
-	if !ok {
-		return ErrResourceNotFound(id.String())
+func AppendManifestToBuffer(manifest []byte, buffer *bytes.Buffer) error {
```

Comment:
The use of YAML doesn't really leak since it's enclosed in the function.

> Can it be worked around by parsing (or otherwise using) the generated manifests separately?

You mean parsing the manifests into a map to transform them into YAML later on? How do you convert them to YAML with the current interface (or a considerable change) without leaking details?

Reply:
> The use of YAML doesn't really leak

Concatenating documents with `---` only works for YAML. So the procedure is specific to YAML documents, but is used from code which makes no assumptions about the serialised form of resources (i.e., what is generated by config files).

> You mean parsing the manifests into a map to transform them into YAML later on?

I meant to parse each chunk of generated config separately, since you can't assume they are YAML. But I'd misread the code as _only_ using the concatenated, generated config as input to `manifests.ParseManifest`. It _also_ uses the result as an argument to `manifests.ApplyManifestPatch`, so that would have to be adapted (or be able to be safely called on each chunk of config separately).

It'll be reasonably easy to troubleshoot or spot ahead of time, I would think, if this turns out to be a problem later. So an alternate mitigation is to be transparent that this assumption is built in, and guard against it being broken accidentally (perhaps by forcing a cast to `YAMLbytes` somewhere? I dunno)

Reply:
Yep, the complicated part are the patching functions.

I tried a type synonym for `[]byte` but it ended up propagating through most of the Manifests interface.